src/Workspaces/Core/Portable/SourceGeneration/IRemoteSourceGenerationService.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Runtime.Serialization;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Text;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.SourceGeneration;
+
+internal interface IRemoteSourceGenerationService
+{
+    /// <summary>
+    /// Given a particular project in the remote solution snapshot, return information about all the generated documents
+    /// in that project.  The information includes the <see cref="SourceGeneratedDocumentIdentity"/> identity
+    /// information about the document, as well as its text <see cref="Checksum"/>.  The local workspace can then
+    /// compare that to the prior generated documents it has to see if it can reuse those directly, or if it needs to
+    /// remove any documents no longer around, add any new documents, or change the contents of any existing documents.
+    /// </summary>
+    ValueTask<ImmutableArray<(SourceGeneratedDocumentIdentity documentIdentity, SourceGeneratedDocumentContentIdentity contentIdentity)>> GetSourceGenerationInfoAsync(
+        Checksum solutionChecksum, ProjectId projectId, CancellationToken cancellationToken);
+
+    /// <summary>
+    /// Given a particular set of generated document ids, returns the fully generated content for those documents.
+    /// Should only be called by the host for documents it does not know about, or documents whose checksum contents are
+    /// different than the last time the document was queried.
+    /// </summary>
+    ValueTask<ImmutableArray<string>> GetContentsAsync(
+        Checksum solutionChecksum, ProjectId projectId, ImmutableArray<DocumentId> documentIds, CancellationToken cancellationToken);
+}
+
+/// <summary>
+/// Information that uniquely identifies the content of a source-generated document and ensures the remote and local
+/// hosts are in agreement on them.
+/// </summary>
+/// <param name="Checksum">Checksum produced from <see cref="SourceText.GetChecksum"/>.</param>
```

Comment:
oh, does it?  let's talk tomorrow.  i want to make sure i'm on the same page as you.

Reply:
so, i don't see a way to avoid this type.  While the text has an encoding/checksum-algo, we still need to *rehydrate* the source-text on the host side, and that def needs this data.  Note: this is similar to what serializablesourcetext has to do as well.